ls1intum/Artemis
# `Programming exercises`: Improve build triggers by instructors

This PR simplifies the process in two ways:
1. Artemis only triggers builds for participation that already include at least one submission: It does not make sense that instructors can trigger student builds when they have not yet submitted something, in fact this could even confuse the student
2. The code does not try to retrieve the latest submission based on the commit hash any more so no failures in doing so cannot break the process. Instead it simply tries to get the last submission from the participation.

There are further changes needed in the web socket communication from server to client (it should be based on the participation entity, not on the submission entity, because triggering a build should not create submissions any more because a submission supports multiple results now), however this should be done in a follow up PR.

Fixes #4463

### Checklist
#### General


#### Server
#### Changes affecting Programming Exercises

### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 Students
- 1 Programming Exercise

1. Login as instructor and create the programming exercise
2. Start the exercise with at least 2 student accounts: at least 1 student should submit changes, at least 1 student should **not** submit anything
3. Login as instructors and open the participation page.
4. Trigger all builds and wait until they finish
5. Only the participations with at least one submission should be triggered and should receive a result

### Review Progress





#### Code Review
#### Manual Tests
